### PR TITLE
[scudo] Separated committed and decommitted entries.

### PR DESCRIPTION
**Pull request by JoshuaMBa**
Initially, the LRU list stored all mapped entries with no distinction between the committed (non-madvise()'d) entries and decommitted (madvise()'d) entries. Now these two types of entries are separated into two lists, allowing future cache logic to branch depending on whether or not entries are committed or decommitted. Furthermore, the retrieval algorithm will prioritize committed entries over decommitted entries. Specifically, valid-fit, committed entries (not necessarily optimal-fit) are retrieved before optimal-fit, decommitted entries.

### CODE REVIEW

File: compiler-rt/lib/scudo/standalone/secondary.h [hunk 1]
```diff
@@ -335,56 +354,70 @@ template <typename Config> class MapAllocatorCache {
...
+      if (!Entry.isValid())
+        Entry = FindAvailableEntry(DECOMMITTED);
+
+      if (!Entry.isValid()) {
+        return false;
+      } else {
+        remove(OptimalFitIndex, OptimalFitListType);
         SuccessfulRetrieves++;
       }
     }
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/secondary.h:420]: nit: 
```
} // ScopedLock L(Mutex);
```

File: compiler-rt/lib/scudo/standalone/secondary.h [hunk 2]
```diff
@@ -335,56 +354,70 @@ template <typename Config> class MapAllocatorCache {
...
+      Entry = FindAvailableEntry(COMMITTED);
+      if (!Entry.isValid())
+        Entry = FindAvailableEntry(DECOMMITTED);
+
+      if (!Entry.isValid()) {
+        return false;
+      } else {
+        remove(OptimalFitIndex, OptimalFitListType);
         SuccessfulRetrieves++;
       }
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/secondary.h:419]:
```suggestion
      if (!Entry.isValid())
        return false;
      
      remove(OptimalFitIndex, OptimalFitListType);
      SuccessfulRetrieves++;
```

File: compiler-rt/lib/scudo/standalone/secondary.h [hunk 3]
```diff
@@ -475,71 +513,88 @@ template <typename Config> class MapAllocatorCache {
...
   }
 
-  void remove(uptr I) REQUIRES(Mutex) {
-    DCHECK(Entries[I].isValid());
-
-    Entries[I].invalidate();
-
-    if (I == LRUHead)
-      LRUHead = Entries[I].Next;
+  void unlink(uptr I, EntryListT ListType) REQUIRES(Mutex) {
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/secondary.h:530]: Can we have some comment about the difference between `unlink` and `remove`?

File: compiler-rt/lib/scudo/standalone/secondary.h (hunk 1 above)
Comment (JoshuaMBa) [reply compiler-rt/lib/scudo/standalone/secondary.h:420]: Should I also add this in the `store()` call. I think the other uses of `Mutex` are missing this comment.

Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/secondary.h:420]: I don't have a strict rule for this. I think the scope is too big to identify the boundary so I want some comment at the end. If you see the same problem, feel free to add it

**Review by cferris1000 — APPROVED** [review]
LGTM

**Comment by nico** [comment]
This triggers thread safety diags: http://45.33.8.238/linux/144530/step_3.txt

**Pull request merged**